### DISCUSSION TITLE
external magnetic field changes with time on it's own!

### DISCUSSION THREAD

**Discussion by nash-host** [epochpic/epoch#440] [General]
Hi EPOCH community,

I was trying to use external magnetic field by defining a constant magnetic field Bz with a fields block. I just defined the fields block and nothing else (no laser or particles) and interestingly the field Bz remains constant (defined magnitude) in the zeroth step and then starts decaying with time later!
The field should have remained constant as there are no perturbations (no laser or particles defined)
![Bz_sdf0](https://user-images.githubusercontent.com/59797203/216305923-6923020a-dd3e-4bbf-88f9-d90fc1633568.png)
![Bz_sdf1](https://user-images.githubusercontent.com/59797203/216305928-58ffa5c8-739a-4553-9274-019eb92466e2.png)
 
I checked by replacing Bz with Bx and Bx seems to be constant throughout as expected.
![Bx_sdf0](https://user-images.githubusercontent.com/59797203/216305812-bc15607e-3e8f-4221-8b56-01feaf255f57.png)
![Bx_sdf5](https://user-images.githubusercontent.com/59797203/216305909-c04ad1c1-3a9c-4e6b-9d9d-c18b9b98407a.png)

Here is my input: [input.txt](https://github.com/Warwick-Plasma/epoch/files/10568084/input.txt)

Please let me know how to define an external magnetic field (Bz) that remains constant unless perturbed by laser or charged particles.

Thank you :)

**Comment by TomGoffrey** [comment]
I don't understand why you think the first field should remain constant with time. Check Maxwell's equations. Your first field has non-zero curl - you'll generate an electric field, which in turn will cause your initial magnetic field to evolve.

As a side note, your second magnetic field is not divergence free.

At its core EPOCH solves Maxwell's equations (whether or not you have a laser or particles present), and I think what you see is consistent with that.

**Reply by nash-host** [reply]
Thank you very much @TomGoffrey  for your response. My bad, I overlooked the basics.
Could you please shed some light on how to define a static magnetic field via input deck in general?

Thank you

**Reply by TomGoffrey** [reply]
The main question here is, what exactly are you trying to model?

As I said, the core solver in EPOCH solves Maxwell's equations - one of which says the magnetic field must be divergence free, and another states that for the Electric field to remain constant (in the absence of a current), requires the magnetic field to be curl free.

So what equations are you trying to solve?

**Reply by nash-host** [reply]
Yes, sure. Many thanks for your help.

I have carried out EPOCH 2D simulations in an effort to reproduce some of our experimental observations on ion acceleration by TNSA from solid targets. There are some interesting features which we think is because of the fields produced at the laser-solid interaction region.

Therefore, I am trying to observe the particle distribution/trajectories in static fields of similar magnitudes to what is observed from our main simulations. The idea is to consider a distribution of particles intercting with external fields and obtain the corresponding particle trajectories, using EPOCH.

As a starting point, I considered a bunch of electrons (using injector block) of a particular energy travelling along +x direction and applied external field Ex in a small region within the simulation box. I could see the electrons getting beautifully reflected by the field. Similarly, I tried to observe the effect of a static magnetic field. Hence, I was enquiring about ways to define a static magnetic field.

**Reply by TomGoffrey** [reply]
The main problem here is that what you want to do (impose a static, non-evolving field) isn't consistent with EPOCH's core solver.

If you just want to solve for the trajectory of electrons in a (piecewise) constant magnetic field - this is analytic. You could hack EPOCH to switch off the field solver, but I don't see the benefit of doing that when there's a known analytic solution.

**Reply by TomGoffrey** [reply]
Just to expand on this, if you -really- want to switch off the field update in EPOCH you can do so by commenting out the calls to `update_eb_fields_half` and `update_eb_fields_final` in epoch2d.F90. Then you've essentially got a particle solver for a non-evolving field. That being said, we don't generally support users using modified versions of the code. Additionally you'd have quite an expensive code for the job you're doing, but it would be a quick hack to get what you want.

**Reply by TomGoffrey** [reply]
One final suggestion - do you need to include the regions of zero magnetic field at all?

**Reply by nash-host** [reply]
Hi Tom, many thanks for your time and valueable suggestions.

Regarding the zero magnetic field - that's a very good idea to omit zero magnetic field regions rather than switching off the field solver. I will try that. I hope it will suffice. Otherwise, as a last option, I will try to switch off the filed update and check, as you said, even I wouldn't prefer modifying the code (as I am a novice user).

Thanks a lot for your help, cheers :)

**Reply by TomGoffrey** [reply]
No problem. Depending on what you're trying to observe it might be worth enabling `ejected_particles` in the output block so you track particles leaving the domain. Or you could consider using the probe output block to track particles going through a certain plane.

**Reply by nash-host** [reply]
yes, I was using probe block till now but I have never tried ejected_particles, I think that will be useful in the case of omiting the zero magnetic field regions. Thanks for the tip.
